includes/widgets/class-wc-widget-layered-nav.php

```diff
@@ -355,34 +355,83 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 			}
 		}
 
-		$meta_query     = new WP_Meta_Query( $meta_query );
-		$tax_query      = new WP_Tax_Query( $tax_query );
-		$meta_query_sql = $meta_query->get_sql( 'post', $wpdb->posts, 'ID' );
-		$tax_query_sql  = $tax_query->get_sql( $wpdb->posts, 'ID' );
-		$term_ids_sql   = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
+		$hide_out_of_stock = 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' );
+		$meta_query        = new WP_Meta_Query( $meta_query );
+		$tax_query         = new WP_Tax_Query( $tax_query );
+		$lookup_table_name = "{$wpdb->prefix}wc_product_attributes_lookup";
+		$meta_query_sql    = $meta_query->get_sql( 'post', $lookup_table_name, 'product_or_parent_id' );
+		$tax_query_sql     = $tax_query->get_sql( $lookup_table_name, 'product_or_parent_id' );
+		$term_ids_sql      = '(' . implode( ',', array_map( 'absint', $term_ids ) ) . ')';
+		$in_stock_clause   = $hide_out_of_stock ? ' AND in_stock = 1' : '';
 
-		// Generate query.
 		$query           = array();
-		$query['select'] = "SELECT COUNT( DISTINCT {$wpdb->posts}.ID ) AS term_count, terms.term_id AS term_count_id";
-		$query['from']   = "FROM {$wpdb->posts}";
+		$query['select'] = 'SELECT COUNT(DISTINCT product_or_parent_id) as term_count, term_id as term_count_id';
+		$query['from']   = "FROM $lookup_table_name";
 		$query['join']   = "
-			INNER JOIN {$wpdb->term_relationships} AS term_relationships ON {$wpdb->posts}.ID = term_relationships.object_id
-			INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy USING( term_taxonomy_id )
-			INNER JOIN {$wpdb->terms} AS terms USING( term_id )
+			INNER JOIN {$wpdb->posts} ON {$wpdb->posts}.ID = {$lookup_table_name}.product_or_parent_id
 			" . $tax_query_sql['join'] . $meta_query_sql['join'];
 
 		$query['where'] = "
 			WHERE {$wpdb->posts}.post_type IN ( 'product' )
 			AND {$wpdb->posts}.post_status = 'publish'
 			{$tax_query_sql['where']} {$meta_query_sql['where']}
-			AND terms.term_id IN $term_ids_sql";
+			AND {$lookup_table_name}.taxonomy='{$taxonomy}'
+			AND {$lookup_table_name}.term_id IN $term_ids_sql
+			{$in_stock_clause}";
+
+		if ( ! empty( $term_ids ) ) {
+			$attributes_to_filter_by = WC_Query::get_layered_nav_chosen_attributes();
+
+			if ( ! empty( $attributes_to_filter_by ) ) {
+				$all_terms_to_filter_by = array();
+				foreach ( $attributes_to_filter_by as $taxonomy => $data ) {
+					$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
+					$term_ids_by_slug           = wp_list_pluck( $all_terms, 'term_id', 'slug' );
+					$term_ids_to_filter_by      = array_values( array_intersect_key( $term_ids_by_slug, array_flip( $data['terms'] ) ) );
+					$all_terms_to_filter_by     = array_merge( $all_terms_to_filter_by, $term_ids_to_filter_by );
```

Comment:
Some very similar logic here that exists in the new `filter_by_attribute_post_clauses` method. Maybe abstract for both?

Also, the same thing applies here on the escaping/sanitization

> I know these are IDs but this should probably be enforcing integers here because they go right into an SQL clause without escaping/sanitization.